add more details for contribution guide (#97)
kgateway-dev/community: CONTRIBUTING.md

## Contributing code

Contributing features to kgateway is a great way to get involved with the project. We welcome contributions of all sizes, from small bug fixes to large new features. Code contribution is done via github's [pull requests](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request) (PRs).

If you are interested in contributing a new feature, please follow these steps:
- For small features (let's say, less than 100 LoC), you can open a pull request (PR) directly.
- For larger features, we recommend opening an issue first to discuss your proposed changes with the community. This will help us understand the problem you are trying to solve, ensure that we are all on the same page, and that your proposed change is aligned with the project.

For large or broad changes, we may ask you to write an enhancement proposal. Use [this template](https://github.com/kgateway-dev/kgateway/blob/main/design/template.md) to get you started.
You can find the existing ehancement proposals [here](https://github.com/kgateway-dev/kgateway/tree/main/design).

> To help you get started with contributing code, We have an [example plugin](https://github.com/kgateway-dev/kgateway/tree/main/examples/plugin) in the kgateway repo. A write-up explaining how plugins work is available [here](https://github.com/kgateway-dev/kgateway/blob/main/devel/architecture/kgateway/DEV.md). We also recomend looking at past PRs that are doing similar things to what you are trying to do. 

Reviewer: Much of this section is already in "Small changes" and "Large changes" below.  Could you have a look and see what isn't (the plugin stuff for example) and merge the two sections to your liking?  I can do this if you prefer. (I do like having a "Contributing code" heading)

Author: thanks! not sure how i missed that

Author: i wanted this section before the "requirements for PRs". So i'll move and merge the  "Ways to contribute" up.

git rebase --signoff main
```

### Testing

> Tests are essential for any non-trivial pull request (PR). They ensure that your changes remain stable and do not break due to future updates. Tests are a critical part of maintaining kgateway's stability and long-term maintainability.

Reviewer: ```suggestion
Tests are essential for any non-trivial PR. They ensure that your feature remains operational and does not break due to future updates. Tests are a critical part of maintaining kgateway's stability and long-term maintainability.
```

When writing tests, consider how users will interact with the feature and design the tests to reflect user behavior - i.e. test what the user would care about.

We have the following types of tests:

#### Unit Tests

> - These are useful for testing small, isolated units of code, such as a single function or a small component.

Reviewer: ```suggestion
These are useful for testing small, isolated units of code, such as a single function or a small component.
```

- Prevents regressions.
- Detects behavior changes from envoy.
- Ensures the feature is not deprecated.
- Confirms the feature works as the user expects it to.

For example, consider this scenario: You are adding a CORS policy. You configure Envoy's route but forget to add the CORS filter. Only an E2E test would catch this issue.

### Error Reporting

If you are contributing a feature that produces error conditions, it is important to surface these errors to the user. Use the following methods to report errors, and try to use **all** of them if possible:
- Logs
- Metrics
- Status on the Custom Resource (using conditions)
- Command-line `check` command (note: this command may not be fully implemented at the time of writing)

If needed, we can also enhance the kgateway admin page with additional supplemental information.

It is important to surface errors in multiple ways because different users interact with kgateway differently. By reporting errors in various places, we maximize the chances of users noticing and addressing them.

### Code review guidelines

> Code may be reviewed by anyone! even if you are not maintainer, please feel free to add your comments.

Reviewer: ```suggestion
Code can be reviewed by anyone! Even if you are not a maintainer, please feel free to add your comments.
```

> All code must be reviewed by at least one [maintainer](https://github.com/kgateway-dev/community/blob/main/MAINTAINERS.md). Key requirements:

Reviewer: ```suggestion
All code must be reviewed by at least one [maintainer](https://github.com/kgateway-dev/community/blob/main/MAINTAINERS.md) before merging. Key requirements:
```

This somewhat undercuts the sentence you added above.  Maybe it can be reviewed by anyone but code must be _approved_ by a maintainer?

git rebase --signoff main
```

> ### Testing

Reviewer 2: as the testing sections below are specific to the `kgateway` repo, should we move it to a readme there instead?

Author: I can see the appeal of moving it there; but i do feel like there should be one place for new contributors to read. the whole PR process is relevant only for `kgateway` repo and were here before...

Author: perhaps this whole doc should move to kgateway repo?
